test/gcs/mock/mock.go

```diff
@@ -0,0 +1,421 @@
+package mock
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"cloud.google.com/go/storage"
+)
+
+// I don't know if it is easier or not to use go mock, but we really only need two things:
+// 1) Ability to mimick creation of buckets and objects
+// 2) Ability to mimick returning errors
+//
+// We don't need arbitrary return values, so generators like go mock or testify might be
+// overkill and doesn't give us the flexibility we need (e.g., will have to specify and
+// and reason about the state after each call rather than just pretend we have this fake
+// storage. The behavior of these commands, at the level of detail we care about, is pretty
+// easy to replicate.
+
+var (
+	MethodNewStorageBucket    = Method("NewStorageBucket")
+	MethodDeleteStorageBucket = Method("NewDeleteStorageBucket")
+	MethodListChildrenFiles   = Method("ListChildrenFiles")
+	MethodListDirectChildren  = Method("ListDirectChildren")
+	MethodAttrObject          = Method("AttrObject")
+	MethodCopyObject          = Method("CopyObject")
+	MethodReadObject          = Method("ReadObject")
+	MethodWriteObject         = Method("WriteObject")
+	MethodDeleteObject        = Method("DeleteObject")
+	MethodDownload            = Method("Download")
+	MethodUpload              = Method("Upload")
+)
+
+// mock GCS Client
+type clientMocker struct {
+	// project with buckets
+	gcp map[project]*buckets
+	// error map
+	// - on each call of the higher level function that calls any number of methods
+	//	in this library, you can use SetError(map[Method]*ReturnError) or ClearError()
+	//	to create the error return values you want. Default is nil.
+	err map[Method]*ReturnError
+
+	// reverse index to lookup which project a bucket is under as GCS has a global
+	// bucket namespace.
+	revIndex map[bucket]project
+}
+
+func NewClientMocker() *clientMocker {
+	c := &clientMocker{
+		gcp:      make(map[project]*buckets),
+		err:      make(map[Method]*ReturnError),
+		revIndex: make(map[bucket]project),
+	}
+	return c
+}
+
+// SetError sets the number of calls of an interface function before an error is returned.
+// Otherwise it will return the err of the mock function itself (which is usually nil).
+func (c *clientMocker) SetError(m map[Method]ReturnError) {
```

Comment:
Should this override?

Reply:
???

Reply:
I am not sure what the question is...

Reply:
I mean `SetError` makes me think replacing the content entirely.
i.e.
```
SetError(key string, err error) {
  c.m[key] = err
}
```
Also
```
SetError(m map[Method]ReturnError) {
  c.m = m
}
```

Reply:
I think from usage point of view, append isn't of a very great use when you are testing. That's why its just set it once when you want to do the test and during the test reset if you need to redefine what errors you are expecting.

Reply:
Oh I see what you are saying...fixed to the second one.